xbmc/platform/posix/filesystem/SMBWSDiscovery.cpp

```diff
@@ -0,0 +1,110 @@
+/*
+ *  Copyright (C) 2021- Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "SMBWSDiscovery.h"
+
+#include "FileItem.h"
+#include "ServiceBroker.h"
+#include "network/IWSDiscovery.h"
+#include "threads/SingleLock.h"
+#include "utils/log.h"
+
+#include "platform/posix/filesystem/SMBWSDiscoveryListener.h"
+
+#include <algorithm>
+#include <chrono>
+#include <memory>
+#include <string>
+#include <vector>
+
+using namespace std::chrono;
+using namespace WSDiscovery;
+
+namespace WSDiscovery
+{
+std::unique_ptr<IWSDiscovery> IWSDiscovery::GetInstance()
+{
+  return std::make_unique<WSDiscovery::CWSDiscoveryPosix>();
+}
+
+CWSDiscoveryPosix::CWSDiscoveryPosix()
+{
+  // Set our wsd_instance ID to seconds since epoch
+  auto epochduration = system_clock::now().time_since_epoch();
+  wsd_instance_id = epochduration.count() * system_clock::period::num / system_clock::period::den;
+
+  m_WSDListenerUDP = std::make_unique<CWSDiscoveryListenerUDP>();
+}
+
+CWSDiscoveryPosix::~CWSDiscoveryPosix()
+{
+  StopServices();
+}
+
+bool CWSDiscoveryPosix::StopServices()
+{
+  CLog::Log(LOGINFO, "CWSDiscoveryPosix::StopServices - Stopping");
+  m_WSDListenerUDP->Stop();
+
+  return true;
+}
+
+bool CWSDiscoveryPosix::StartServices()
+{
+  CLog::Log(LOGINFO, "CWSDiscoveryPosix::StartServices - Starting");
+  m_WSDListenerUDP->Start();
+
+  return true;
+}
+
+bool CWSDiscoveryPosix::IsRunning()
+{
+  return m_WSDListenerUDP->IsRunning();
+}
+
+bool CWSDiscoveryPosix::GetServerList(CFileItemList& items)
+{
+  {
+    CSingleLock lock(m_critWSD);
+
+    // delim1 used to strip protocol from xaddrs
+    // delim2 used to strip anything past the port
+    const std::string delim1 = "://";
+    const std::string delim2 = ":5357";
```

Comment:
Unicast SOAP HTTP WS-Dicsovery responder are allowed to listen on any TCP port, e.g. wsdd2 is using 3702.

Reply:
Can you show me a log for an instance of this?
This was very much specific to what I could test (win10, Synology wsd), so if you can give me a log so I can see how I need to handle the xaddrs value, that'd be great.

Reply:
Log when using the "https:" actions: http://ix.io/3ulQ. Neither the router nor the wsdd2 device is listed.

With "http:" actions both devices are detected: http://ix.io/3ulS. Wsdd2 is using 192.168.11.3:3702

Reply:
Cheers, i'll look to generalise the xaddrs stripping a bit more hopefully this weekend.
I'll work out a way to handle the actions as well. According to the spec the "actions" can effectively be disregarded, which might be why windows and wsd.py implementations might not care about what is sent for that part of the SOAP xml packet, and hence they work with this implementation.

Reply:
Have altered the delimiter matching to match on the colon instead of the :5357. Id expect it to be ok as that, but give it a test and let me know how it goes.

As part of this ive also moved the WS-Discovery logging to a new component logging. I figure for the most part we dont really care about WS-D unless we want to debug it, so a user can turn on component logging to get the WS-D debug stuff.
With that, ive added some extra logging in states to provide some info about the received packets that are discarded for various reasons. Should give more info in case people come across other setups in the future.

Reply:
Looks good. All devices are visible now.

Reply:
Thanks for testing, greatly appreciated.